coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,181 @@
+	tickCh   <-chan time.Time       // controls flush interval
+	stopTick func()                 // stops flushing
+	stopCh   chan struct{}          // signals us to stop
+	stopOnce sync.Once              // because you only stop once
+	doneCh   chan struct{}          // signifies that we have stopped
+	flushCh  chan int               // used for testing.
+}
+
+// New returns a new Tracker. It is the caller's responsibility
+// to call Close().
+func New(s Store, opts ...Option) *Tracker {
+	hb := &Tracker{
+		log:      slog.Make(sloghuman.Sink(os.Stderr)),
+		m:        make(map[uuid.UUID]struct{}, 0),
+		s:        s,
+		tickCh:   nil,
+		stopTick: nil,
+		stopCh:   make(chan struct{}),
+		doneCh:   make(chan struct{}),
+		flushCh:  nil,
+	}
+	for _, opt := range opts {
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only.
+func WithFlushChannel(c chan int) Option {
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only.
+func WithTickChannel(c chan time.Time) Option {
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.mut.Lock()
+	wut.m[workspaceID] = struct{}{}
+	wut.mut.Unlock()
+}
+
+// flushLocked updates last_used_at of all current workspace IDs.
+// MUST HOLD LOCK BEFORE CALLING
+func (wut *Tracker) flushLocked(now time.Time) {
+	if wut.mut.TryLock() {
+		panic("developer error: must lock before calling flush()")
+	}
+	count := len(wut.m)
+	defer func() { // only used for testing
+		if wut.flushCh != nil {
+			wut.flushCh <- count
+		}
+	}()
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+	// Copy our current set of IDs
+	ids := make([]uuid.UUID, 0)
+	for k := range wut.m {
+		ids = append(ids, k)
+	}
+	// Reset our internal map
+	wut.m = make(map[uuid.UUID]struct{})
+	// For ease of testing, sort the IDs lexically
+	sort.Slice(ids, func(i, j int) bool {
+		// For some unfathomable reason, byte arrays are not comparable?
+		return strings.Compare(ids[i].String(), ids[j].String()) < 0
+	})
```

Comment:
self-review: This makes me sad and I would love to know a better way to do this.
The main reason I'm using a `map[uuid.UUID]struct{}` here is to de-duplicate.

Reply:
The only thing I can think of is to maintain a slice and a map in parallel. In the channel method I mentioned, both would be function scoped variables.

In practice though, sorting a couple thousand IDs once per minute is not a big deal imo.

Reply:
for the string thing, you could do:
```golang
	sort.Slice(ids, func(i, j int) bool {
		return bytes.Compare(ids[i][:], ids[j][:]) < 0
	})
```

Reply:
🧠